telemetry/src/metrics.rs

```diff
@@ -0,0 +1,174 @@
+    pub isi: IntCounterVec,
+    /// Query handle time Histogram
+    pub isi_times: HistogramVec,
+    // Internal use only.
+    registry: Registry,
+}
+
+impl Default for Metrics {
+    // The constructors either always fail, or never.
+    #[allow(clippy::expect_used)]
+    fn default() -> Self {
+        let txs = IntCounterVec::new(Opts::new("txs", "Transactions committed"), &["type"])
+            .expect("Infallible");
+        let isi = IntCounterVec::new(
+            Opts::new("isi", "Iroha special instructions handled by this peer"),
+            &["type", "success_status"],
+        )
+        .expect("Infallible");
+        let isi_times = HistogramVec::new(
+            HistogramOpts::new("isi_times", "Time to handle isi in this peer"),
+            &["type"],
+        )
+        .expect("Infallible");
+        let tx_amounts = Histogram::with_opts(HistogramOpts::new(
+            "tx_amount",
+            "average amount involved in a transaction on this peer",
+        ))
+        .expect("Infallible");
+        let block_height =
+            IntCounter::new("block_height", "Current block height").expect("Infallible");
+        let connected_peers = GenericGauge::new(
+            "connected_peers",
+            "Total number of currently connected peers",
+        )
+        .expect("Infallible");
+        let uptime_since_genesis_ms = GenericGauge::new(
+            "uptime_since_genesis_ms",
+            "Network up-time, from creation of the genesis block",
+        )
+        .expect("Infallible");
+        let domains = GenericGauge::new("domains", "Total number of domains").expect("Infallible");
+        let accounts = GenericGaugeVec::new(
+            Opts::new("accounts", "User accounts registered at this time"),
+            &["domain"],
+        )
+        .expect("Infallible");
+        let registry = Registry::new();
+
+        macro_rules! register {
+			($metric:expr)=> {
+				registry.register(Box::new($metric.clone())).expect("Infallible");
+			};
+			($metric:expr,$($metrics:expr),+)=>{
+				register!($metric);
+				register!($($metrics),+);
+			}
+		}
+
+        register!(
+            txs,
+            tx_amounts,
+            block_height,
+            connected_peers,
+            uptime_since_genesis_ms,
+            domains,
+            accounts,
+            isi,
+            isi_times
+        );
+
+        Self {
+            txs,
+            block_height,
+            connected_peers,
+            uptime_since_genesis_ms,
+            registry,
+            domains,
+            accounts,
+            tx_amounts,
+            isi,
+            isi_times,
+        }
+    }
+}
+
+impl Metrics {
+    /// Convert the current [`Metrics`] into a Prometheus-readable format.
+    ///
+    /// # Errors
+    /// - If [`Encoder`] fails to encode the data
+    /// - If the buffer produced by [`Encoder`] causes [`String::from_utf8`] to fail.
+    pub fn try_to_string(&self) -> eyre::Result<String> {
```

Comment:
would it make sense to use [this](https://docs.rs/serde_prometheus/latest/serde_prometheus/)? This **is** a form of serialization

Reply:
it may not be a good idea to depend on this lib, but rather implement `Deserialize` on metrics by hand

Reply:
I tried to keep it minimal. `prometheus` has very few dependencies, so if we wanted to add some metrics to a leaf crate, it doesn't immediately need to depend on `serde`. I do think that maybe using this in `core` may be a good idea.

Reply:
yeah, I agree, let's keep dependencies to minimum. I've seen some libraries expose `serde` feature to keep dependencies minimal

Reply:
Ah. I understand now. The issue with trying to do that has to do with the histogram metrics. It would be a lot of work, which would be better reserved for a separate PR. Believe me, having many issues resolved in one PR isn't easy to review or merge.